app/models/concerns/client_search.rb

```diff
@@ -44,36 +30,47 @@ def self.text_searcher(text, **kwargs, &block)
       elsif numeric
         where = sa[:PersonalID].eq(text).or(sa[:id].eq(text))
       else
-        query = "%#{text.downcase}%"
-        where = name_search(sa, :FirstName, query, **kwargs).
-          or(name_search(sa, :LastName, query, **kwargs))
+        matches_external_ids = false
+        if alpha_numeric && respond_to?(:search_by_external_id) && RailsDrivers.loaded.include?(:hmis_external_apis)
+          eid_t = HmisExternalApis::ExternalId.arel_table
+          external_id_scope = HmisExternalApis::ExternalId.for_clients
+            .where(eid_t[:value].eq(text))
+            .select(:source_id)
+          matches_external_ids = self.where(id: external_id_scope.select(:source_id)).any?
+        end
```

Comment:
Why is this needed in addition to the `search_by_external_id` that is already called on line 51?

It doesnt seem like the right place to call it. If the external ID is numeric, or alphanumeric with length 32 or 36, this code wont get hit.

Reply:
So my thought was that it could be hard to combine ranked fuzzy name search with these other search conditions inside a single query, which is how the other conditions are handled. Perhaps there's an easy way to do that but it also could be a time sink.

To avoid dealing with it, we do an early return if there are no external id matches (at the expense of the extra query). Perhaps there's a better way to differentiate ID vs name search. Maybe if the term contains no numeric chars, we treat it as a name, otherwise it's a ID? Could also do the name search and return early if there are any results (assuming an ID would never fuzzy-match to a name).

> It doesnt seem like the right place to call it. If the external ID is numeric, or alphanumeric with length 32 or 36, this code wont get hit.

We shouldn't need to differentiate id vs name search if the term is numeric. However the alphanumeric string with 'magic' 32/36 lengths seems like it would be incorrect for searches against very long human names. That's the current behavior though.

Happy to huddle about this if needed.

Reply:
Ok, I think I get it. This extra query is here because we want to _avoid_ doing the NameSearch if the search query matches an external ID.

If the search query is numeric, the External ID query still occurs, it just happens down on line 55. 

Did I get that right?  Can we add a comment about why we're doing the search twice?

All of the External Ids we have right now are numeric, so `matches_external_ids` will never be true. But I can see the utility for the future, if we have some _alphanumeric_ external IDs to support.


Finally - do we need to repeat the actual query logic for searching by external ID? Can this be replaced with the below?

```
empty_where_clause = sa[:id].eq(nil)
matches_external_ids  = search_by_external_id(empty_where_clause, text).exists?
```


And - You're right, the 32/36 magic length would mess with searching for a name of that length. But to be considered "alphanumeric" there wouldn't be any spaces.. I think thats uncommon enough we should leave that behavior as-is.

Reply:
Sounds good, thanks @ttoomey. Commenting it out for now seems like a good call.